_source/_posts/2022-01-05-auth-patterns-php-microservices.md

```diff
@@ -0,0 +1,576 @@
+
+This tutorial also heavily utilizes Okta for the various authentication patterns you will implement. As such, you will need to create an [Okta Developer account](https://developer.okta.com/signup). For now, create a free account. The tutorial will cover any operations you need to perform as they come up later.
+
+Finally, for testing the APIs and authentication methods, make sure you have the latest version of [Postman](https://www.postman.com/downloads/) installed. Postman dramatically simplifies the process of getting OAuth tokens and calling the APIs.
+
+As this tutorial primarily focuses on the authentication patterns themselves, the actual microservices are not overly significant as long as you have an API to call. You can create some [Laravel](https://laravel.com/) projects to get a few microservices up and running. Though Laravel is a bit on the heavy side for such simple use cases, it is quick and easy to set up, comes with most of the libraries you will need out-of-the-box, and has a sound middleware system that you can use to implement these auth patterns. To create the Laravel projects, the official documentation suggests using the `laravel.build` site, which serves as a Bash script for you to run. It is a good idea to read the script yourself before piping it into Bash, so be sure to [review the script](https://laravel.build/api-gateway?with=redis) before executing it if you are uncertain. It works by taking the `path` parameter and using that as the project name.
+
+### Getting started with microservices
+
+To begin, create a directory to hold all of the microservices for this tutorial named something like `php-microservices-demo`. `cd` into this directory, and execute the following command: `curl -s "https://laravel.build/api-gateway?with=redis" | bash`. This will create a directory containing a Laravel project named `api-gateway`. Once this step is done, you will need to run it four more times like so; once for each of the microservices you will create:
+```bash
+curl -s "https://laravel.build/microservice-a?with=redis" | bash
+curl -s "https://laravel.build/microservice-b?with=redis" | bash
+curl -s "https://laravel.build/microservice-c?with=redis" | bash
+curl -s "https://laravel.build/microservice-d?with=redis" | bash
+Now our directory should contain five projects in total. These projects are intended to run with [Laravel Sail](https://laravel.com/docs/8.x/sail), but it isn’t optimal for running five of them simultaneously. Use Sail to install dependencies and execute `artisan` commands, but for running the APIs, create a new `docker-compose.yml` file in the tutorial directory. This file will allow you to start all the containers together. Luckily, because the APIs will be simple, you don’t even need databases for this, only the PHP containers. In your newly created `docker-compose.yml`, add the following content:
```

Comment:
```suggestion
To begin, create a directory to hold all of the microservices for this tutorial named something like `php-microservices-demo`. `cd` into this directory, and execute the following command: `curl -s "https://laravel.build/api-gateway?with=redis" | bash`. This will create a directory containing a Laravel project named `api-gateway`. Once this step is done, you will need to run it four more times like so; once for each of the microservices you will create:

```bash
curl -s "https://laravel.build/microservice-a?with=redis" | bash
curl -s "https://laravel.build/microservice-b?with=redis" | bash
curl -s "https://laravel.build/microservice-c?with=redis" | bash
curl -s "https://laravel.build/microservice-d?with=redis" | bash


Now our directory should contain five projects in total. These projects are intended to run with [Laravel Sail](https://laravel.com/docs/8.x/sail), but it isn’t optimal for running five of them simultaneously. Use Sail to install dependencies and execute `artisan` commands, but for running the APIs, create a new `docker-compose.yml` file in the tutorial directory. This file will allow you to start all the containers together. Luckily, because the APIs will be simple, you don’t even need databases for this, only the PHP containers. In your newly created `docker-compose.yml`, add the following content:
```